frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+//! lists accordingly. It also exposes [`TrackedList`] that adds defensive checks to a subset of
+//! [`SortedListProvider`] methods in order to spot unexpected list updates on the consumer side.
+//! This wrapper should beused to pass the tracked entity to the consumer.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+pub(crate) mod mock;
+#[cfg(test)]
+mod tests;
+
+use frame_election_provider_support::{SortedListProvider, VoteWeight};
+use frame_support::{
+	defensive,
+	traits::{Currency, CurrencyToVote},
+};
+pub use pallet::*;
+use sp_runtime::{
+	traits::{Bounded, Zero},
+	Saturating,
+};
+use sp_staking::{OnStakingUpdate, Stake, StakingInterface};
+
+use sp_std::{boxed::Box, vec::Vec};
+
+/// The balance type of this pallet.
+pub type BalanceOf<T> = <<T as Config>::Staking as StakingInterface>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::*;
+	use frame_election_provider_support::{SortedListProvider, VoteWeight};
+	use frame_support::pallet_prelude::*;
+
+	use sp_staking::StakingInterface;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// This has to come from Staking::Currency
+		type Currency: Currency<Self::AccountId, Balance = BalanceOf<Self>>;
+
+		type Staking: StakingInterface<AccountId = Self::AccountId>;
+
+		/// A sorted list of nominators and validators, by their stake and self-stake respectively.
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn active_stake_of(who: &T::AccountId) -> BalanceOf<T> {
+		T::Staking::stake(&who).map(|s| s.active).unwrap_or_default()
+	}
+
+	pub(crate) fn to_vote(balance: BalanceOf<T>) -> VoteWeight {
+		let total_issuance = T::Currency::total_issuance();
+		<T::Staking as StakingInterface>::CurrencyToVote::to_vote(balance, total_issuance)
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	fn on_stake_update(who: &T::AccountId, _: Option<Stake<T::AccountId, BalanceOf<T>>>) {
+		if let Ok(current_stake) = T::Staking::stake(who) {
+			let current_active = current_stake.active;
+
+			// if this is a nominator
+			if let Some(_) = T::Staking::nominations(&current_stake.stash) {
+				let _ =
+					T::VoterList::on_update(&current_stake.stash, Self::to_vote(current_active));
+			}
+
+			if T::Staking::is_validator(&current_stake.stash) {
+				let _ =
+					T::VoterList::on_update(&current_stake.stash, Self::to_vote(current_active));
+			}
+		}
+	}
+
+	fn on_nominator_update(who: &T::AccountId, _prev_nominations: Vec<T::AccountId>) {
+		// NOTE: We ignore the result here, because this method can be called when the nominator
```

Comment:
It looks like it's better to have an `on_nominator_insert` and call if the Nominator wasn't previously in the list. Seems legit to indeed decouple insert and update logic to avoid confusion. A bit more verbose though.